grype/pkg/syft_json_provider.go

```diff
@@ -0,0 +1,213 @@
+package pkg
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"strings"
+
+	"github.com/anchore/grype/grype/cpe"
+	"github.com/anchore/syft/syft/distro"
+	"github.com/anchore/syft/syft/pkg"
+	syftJson "github.com/anchore/syft/syft/presenter/json"
+	"github.com/anchore/syft/syft/source"
+)
+
+func syftJSONProvider(config providerConfig) ([]Package, Context, error) {
+	var reader io.Reader
+	if config.reader != nil {
+		// the caller is explicitly hinting to use the given reader as input
+		reader = config.reader
+	} else {
+		// try and get a reader from the description
+		if strings.HasPrefix(config.userInput, "sbom:") {
+			// the user has explicitly hinted this is an sbom, if there is an issue return the error
+			filepath := strings.TrimPrefix(config.userInput, "sbom:")
+			sbomReader, err := os.Open(filepath)
+			if err != nil {
+				return nil, Context{}, fmt.Errorf("user hinted 'sbom:' but could read SBOM file: %w", err)
+			}
+			reader = sbomReader
+		}
+
+		// the user has not hinted that this may be a sbom, but lets try that first...
+		if sbomReader, err := os.Open(config.userInput); err != nil {
+			if errors.Is(err, os.ErrNotExist) {
+				return nil, Context{}, errDoesNotProvide
+			}
+		} else {
+			reader = sbomReader
+		}
+	}
+
+	return parseSyftJSON(reader)
+}
+
+// partialSyftDoc is the final package shape for a select elements from a syft JSON document.
+type partialSyftDoc struct {
+	Source    syftJson.Source       `json:"source"`
+	Artifacts []partialSyftPackage  `json:"artifacts"`
+	Distro    syftJson.Distribution `json:"distro"`
+}
+
+// partialSyftPackage is the final package shape for a select elements from a syft JSON package.
+type partialSyftPackage struct {
+	packageBasicMetadata
+	packageCustomMetadata
+}
+
+// packageBasicMetadata contains non-ambiguous values (type-wise) from pkg.Package.
+type packageBasicMetadata struct {
+	Name      string            `json:"name"`
+	Version   string            `json:"version"`
+	Type      pkg.Type          `json:"type"`
+	Locations []source.Location `json:"locations"`
+	Licenses  []string          `json:"licenses"`
+	Language  pkg.Language      `json:"language"`
+	CPEs      []string          `json:"cpes"`
+	PURL      string            `json:"purl"`
+}
+
+// packageCustomMetadata contains ambiguous values (type-wise) from pkg.Package.
+type packageCustomMetadata struct {
+	MetadataType pkg.MetadataType `json:"metadataType"`
+	Metadata     interface{}      `json:"metadata"`
+}
+
+// packageMetadataUnpacker is all values needed from Package to disambiguate ambiguous fields during json unmarshaling.
+type packageMetadataUnpacker struct {
+	MetadataType pkg.MetadataType `json:"metadataType"`
+	Metadata     json.RawMessage  `json:"metadata"`
+}
+
+// JavaMetadata encapsulates all Java ecosystem metadata for a package as well as an (optional) parent relationship.
```

Comment:
(nit) this and the following doc comments won't be picked up by dev tooling since the names don't match exactly

Reply:
yup, copy-pasta problem, I'll adjust